### DISCUSSION TITLE
entproto json support

### DISCUSSION THREAD

**Discussion by tarrencev** [ent/ent#1690] [Ideas]
It would be great to support json fields in entproto. Ideally, they can be marshalled into a proto themselves. My proposal would be to so similar to [gqlgen](https://gqlgen.com/reference/scalars/#custom-scalars-with-user-defined-types) user defined types. We can support a `MarshalProto` and `UnmarshalProto` interface, which will marshal/unmarshal the field to/from a custom defined proto. In this case, the proto wouldn't be generated by entproto.

**Comment by rotemtam** [comment]
I think supporting JSON fields would be really great. I'm not sure I understand how to map the gqlgen style API into ent + entproto + the generated gRPC code.  Can you provide some psuedo code on what that would look like? 

Thanks 🙏

**Comment by tarrencev** [comment]
Now that we have merged https://github.com/ent/contrib/pull/96, I think we can do something simpler.

Something like this:
```go
func (Proposal) Fields() []ent.Field {
	return []ent.Field{
		field.JSON("executable", &Executable{}).
			Immutable().
			Annotations(
				entproto.Field(1),
			),
	}
}

type Calldata struct {
	Data []byte `json:"data,omitempty", entproto:"data,1"`
}

type Executable struct {
	Targets    []common.Address `json:"targets,omitempty", entproto:"targets,1"`
	Values     []*big.Int       `json:"values,omitempty", entproto:"values,2"`
	Signatures []string         `json:"signatures,omitempty", entproto:"signatures,3"`
	Calldatas  []Calldata         `json:"calldatas,omitempty", entproto:"calldatas,4"`
}
```

Would generate something like:
```
message Calldata {
	repeated bytes data = 1;
}

message Executable {
	repeated bytes targets = 1;
	repeated bytes values = 2;
	repeated string signatures = 3;
	repeated Calldata calldatas = 4;
}
```

For custom GoTypes, they must implement the [BinaryMarshaller](https://golang.org/pkg/encoding/#BinaryMarshaler)/[BinaryUnmarshaller](https://golang.org/pkg/encoding/#BinaryUnmarshaler) interface.

**Reply by rotemtam** [reply]
I like this direction, but it kind of sucks to have two kinds of syntax for describing the field -> number relations. 
In addition, annotations are serialized with the rest of the schema which may be helpful in the future. 

Not sure about it but what if we used syntax similar to Enums:
```go
func (Proposal) Fields() []ent.Field {
	return []ent.Field{
		field.JSON("executable", &Executable{}).
			Immutable().
			Annotations(
				entproto.Field(1),
				entproto.JSON(map[string]int32{
					"targets": 1,
					"values": 2,
					//...
				}),
	}
}
```
This has some benefits as mentioned, but not sure how we would describe nested structs.

Maybe the direction should be that for JSON fields, the expectation is that the pb is defined elsewhere, and an annotation referencing that already existing type is provided?

**Reply by tarrencev** [reply]
Yeah that is the biggest issue, how to handle nested structs.

I know this is a big change i direction, but we could modify entproto to only use struct fields.

So instead of `entproto` annotations, we would do something like

```
field.JSON("executable", &Executable{}).
            StructTag(`json:"executable,omitempty", entproto:"executable,1"`),
```

**Reply by rotemtam** [reply]
I understand your suggestion and I see how it will provide a unified API but it's very different from how everything works in Ent. One of the reasons (I think) users love ent is the explicit, type-safe API and I'd really like to avoid struct tags which personally find much harder to use (I know they are very common in Go, but I always get into trouble with them.)

**Reply by rotemtam** [reply]
I think that a direction could be that the pb is assumed to be defined elsewhere, and the field definition would be something similar to:
```go
field.JSON("executable", &Executable{}).
  .Annotations(
    entproto.Field(10),
    entproto.Custom( /* some descriptor */),
  )
```
Hopefully, during the `.proto` file generation, we can be smart enough to create a pb file that references this type (or if we cant replicate it).

In addition, we need to figure out the facilities for going from the JSON struct to the  proto struct and vice versa. Maybe they can be the same?

**Reply by tarrencev** [reply]
Yeah totally understand. I think your proposed solution would work well too

**Comment by pavelsmejkal** [comment]
I Would like to contribute to this discussion with a proposal based on https://github.com/ent/contrib/pull/213 and @rotemtam comment/idea. 
I would prefer to have granular control over generated proto. To be able to support existing systems with already defined messages or API. Allow inc

I can see several phases that could bring entproto closer to the desired state. 

1. Template-based proto generation
2. Support for field name override
3. Named messages


#### Template based proto generation

In order to get full control over a generated proto message, we might derive the final proto descriptor based given template, respectively based on the descriptor set. 
A template might be defined per package. In order to provide entproto templates, we can define global annotation supplied by gen.Config. 
There are two options to how to inject the global annotation into entproto. 

1. Adding new flag -config into entproto bin with json,yaml,tomp,what ever format and unmarshal it into gen.Config. 
2. Use a custom instance of entproto copied into the project with custom config values. 

Example of option 2)
```
	graph, err := entc.LoadGraph(*schemaPath, &gen.Config{
		Target: "./foo",
		Annotations: gen.Annotations{}.Add(
			entproto.Packages(
				entproto.Package("entpb").Template("templates/message.protoset"),
			),
		),
	if err != nil {
		log.Fatalf("entproto: failed loading ent graph: %v", err)
	}
	if err := entproto.Generate(graph); err != nil {
		log.Fatalf("entproto: failed generating protos: %s", err)
	}
		
```

The template could be defined using protoset. Later on the system can try to autogenerate protoset based on the given proto message. 
The descriptor set can be easily derived from the template using:
```
//go:generate protoc --descriptor_set_out=./message.protoset --include_imports -I. message.proto
```

The generation process would prioritize descriptors already existing in the template. 

- Message/Services deduped by entity name
- Fields are deduped based on proto field numbers to let users define custom name and type. 

To signal GRPC service generator that some fields require custom handling we might expose custom field option extension. 


```
string field_name = 2 [(entproto.field) = { custom:true }];
or 
string field_name = 2 [(entproto.field) = { binding:false }];
```


#### Support for field name override

In order to be able to migrate already existing API, entproto could allow specifying custom field name that will be used in proto message.
I know that proto guidelines talk about snake case fields but there are very diverse API that needs to be maintained. 

```
field.String("my_name").Annotations(
  entproto.Field(2, entproto.FieldName("MyName")),
),

```
Result: 

```
string MyName = 2;
```

#### Named messages
The way how to generate scenario-based proto messages based on particular fields. For example for API that allows updates on specific fields only. 
This proposal will be covered separately since it might be big bigger.

**Reply by rotemtam** [reply]
hey @pavelsmejkal 
great seeing you here :-)

I will respond regarding your ideas template-based and named messages a bit later, but regrading field name overrides, I think it is a reasonable expectation to allow users to configure the field name. If you would like to contribute this change I will gladly work with you on it 💪
